packages/sdk/src/protocols/filter/reliability_monitor.ts

```diff
@@ -0,0 +1,104 @@
+import type { Peer, PeerId } from "@libp2p/interface";
+import { IProtoMessage, PeerIdStr, PubsubTopic } from "@waku/interfaces";
+import { messageHashStr } from "@waku/message-hash";
+import { WakuMessage } from "@waku/proto";
+import { Logger } from "@waku/utils";
+
+type ReceivedMessageHashes = {
+  all: Set<string>;
+  nodes: Record<PeerIdStr, Set<string>>;
+};
+
+const DEFAULT_MAX_MISSED_MESSAGES_THRESHOLD = 3;
+
+const log = new Logger("sdk:filter:reliability_monitor");
+
+export class ReliabilityMonitor {
+  private receivedMessagesHashes: ReceivedMessageHashes;
+  private missedMessagesByPeer: Map<PeerIdStr, number> = new Map();
+  private maxMissedMessagesThreshold = DEFAULT_MAX_MISSED_MESSAGES_THRESHOLD;
+
+  public constructor(
+    private getPeers: () => Peer[],
+    private renewAndSubscribePeer: (peerId: PeerId) => Promise<Peer | undefined>
+  ) {
+    const allPeerIdStr = this.getPeers().map((p) => p.id.toString());
+
+    this.receivedMessagesHashes = {
+      all: new Set(),
+      nodes: Object.fromEntries(
+        allPeerIdStr.map((peerId) => [peerId, new Set()])
+      )
+    };
+    allPeerIdStr.forEach((peerId) => this.missedMessagesByPeer.set(peerId, 0));
+  }
+
+  public setMaxMissedMessagesThreshold(value: number | undefined): void {
+    if (value !== undefined) {
+      this.maxMissedMessagesThreshold = value;
+    }
+  }
+
+  public addMessage(
+    message: WakuMessage,
+    pubsubTopic: PubsubTopic,
+    peerIdStr?: PeerIdStr
+  ): boolean {
+    const hashedMessageStr = messageHashStr(
+      pubsubTopic,
+      message as IProtoMessage
+    );
+
+    const isNewMessage = !this.receivedMessagesHashes.all.has(hashedMessageStr);
+    this.receivedMessagesHashes.all.add(hashedMessageStr);
+
+    if (peerIdStr) {
+      this.receivedMessagesHashes.nodes[peerIdStr].add(hashedMessageStr);
+    }
+
+    return !isNewMessage;
+  }
+
+  public async validateMessage(): Promise<void> {
+    for (const hash of this.receivedMessagesHashes.all) {
+      for (const [peerIdStr, hashes] of Object.entries(
+        this.receivedMessagesHashes.nodes
+      )) {
+        if (!hashes.has(hash)) {
+          this.incrementMissedMessageCount(peerIdStr);
+          if (this.shouldRenewPeer(peerIdStr)) {
+            await this.renewPeer(peerIdStr);
+          }
+        }
+      }
+    }
+  }
+
+  private incrementMissedMessageCount(peerIdStr: PeerIdStr): void {
+    const currentCount = this.missedMessagesByPeer.get(peerIdStr) || 0;
+    this.missedMessagesByPeer.set(peerIdStr, currentCount + 1);
+  }
+
+  private shouldRenewPeer(peerIdStr: PeerIdStr): boolean {
+    const missedMessages = this.missedMessagesByPeer.get(peerIdStr) || 0;
+    return missedMessages > this.maxMissedMessagesThreshold;
+  }
+
+  private async renewPeer(peerIdStr: PeerIdStr): Promise<void> {
+    log.info(`Peer ${peerIdStr} has missed too many messages, renewing.`);
```

Comment:
I am not sure we need to create wrapper around `this.renewAndSubscribePeer` as it should handle the cases of peer being absent and error handling.

Reply:
in that case I'd use it directly, we already have logging inside